### PR TITLE
feat(comm): add fused decode CP A2A + LSE reduce

### LINKED ISSUE #4575: [Feature] Fuse LSE correction into DCP A2A
## Motivation

SGLang uses FlashInfer's MNNVL DCP A2A operator to exchange partial attention outputs and LSE values across context-parallel ranks. The current API returns the peer-stacked tensors, after which SGLang launches a separate kernel to apply the LSE correction:

```python
recv_o, recv_stats = decode_cp_a2a_alltoall(
    partial_o,
    softmax_stats,
    workspace,
    cp_rank,
    cp_size,
)
output = lse_reduce(recv_o, recv_stats[..., 0])
```

Could FlashInfer provide a fused A2A + LSE-reduction operation that directly returns the corrected attention output?

## Proposed Interface

```python
output = decode_cp_a2a_lse_reduce(
    partial_o,            # [..., cp_size, head_dim], fp16/bf16
    partial_lse,          # [..., cp_size], fp32
    workspace,
    cp_rank,
    cp_size,
    is_lse_base_on_e=False,
    enable_pdl=None,
)
```

## Expected behavior

- Perform the same MNNVL exchange as `decode_cp_a2a_alltoall`.
- Apply the exact LSE-weighted reduction over contributions from all CP ranks.
- Return `output` with shape `[..., head_dim]`, with the CP dimension removed.
- Support both base-2 LSE, used by FlashInfer MLA, and natural-log LSE.
- Remain compatible with CUDA graph capture and replay.
- Ideally avoid materializing peer-stacked receive tensors and launching a separate correction kernel.

For inputs shaped as follows:

```text
partial_o:   [batch, local_heads, cp_size, head_dim]
partial_lse: [batch, local_heads, cp_size]
```

the result should have shape:

```text
output:      [batch, local_heads, head_dim]
```

## Reference validation

The fused result should match the current A2A followed by an explicit LSE reduction:

```python
softmax_stats = torch.zeros(
    *partial_lse.shape, 2, dtype=torch.float32, device=partial_lse.device
)
softmax_stats[..., 0] = partial_lse

recv_o, recv_stats = decode_cp_a2a_alltoall(
    partial_o,
    softmax_stats,
    workspace,
    cp_rank,
    cp_size,
)

recv_lse = recv_stats[..., 0]
lse_max = recv_lse.max(dim=-1, keepdim=True).values
weights = (
    torch.exp(recv_lse - lse_max)
    if is_lse_base_on_e
    else torch.exp2(recv_lse - lse_max)
)
expected = (
    (recv_o.float() * weights.unsqueeze(-1)).sum(dim=-2)
    / weights.sum(dim=-1, keepdim=True)
).to(partial_o.dtype)

actual = decode_cp_a2a_lse_reduce(
    partial_o,
    partial_lse,
    workspace,
    cp_rank,
    cp_size,
    is_lse_base_on_e=is_lse_base_on_e,
)

torch.testing.assert_close(actual, expected, rtol=1e-2, atol=1e-3)
```

### PR DESCRIPTION
**Pull request by kwen2501**
### Why this is needed

In context-parallel decode, the KV cache is sharded by sequence across ranks. Each rank runs attention against
its local KV shard and produces, for every query token and head:

- `o_r`: a partial attention output
- `lse_r`: the log-sum-exp that records that shard's softmax normalization

The `o_r` values cannot be summed directly: each was normalized over a different KV shard. To reconstruct
full-context attention, the partial outputs must be exchanged and merged with weights derived from all ranks'
LSEs:

`out = Σ_r softmax(lse)_r * o_r`

Each rank's input carries a `cp_size` axis containing the slices destined for the CP ranks. The `heads` axis may
contain either local or total heads: this op treats every batch/head entry independently and does not shard or
otherwise reinterpret the head axis.

### What this PR adds

Adds experimental `decode_cp_a2a_lse_reduce`, which fuses that all-to-all exchange and LSE-weighted merge into one
cooperative CUDA kernel.

```python
decode_cp_a2a_lse_reduce(
    partial_o: torch.Tensor,
    partial_lse: torch.Tensor,
    workspace: torch.Tensor,
    cp_rank: int,
    cp_size: int,
    lse_mode: Literal["base2", "basee"] = "base2",
) -> torch.Tensor
```

Input and output shapes:

- `partial_o`: `[batch, heads, cp_size, head_dim]` (`float16` or `bfloat16`)
- `partial_lse`: `[batch, heads, cp_size]` (`float32`)
- output: `[batch, heads, head_dim]`, with the same dtype as `partial_o`

More generally, arbitrary matching leading dimensions are supported:
`[..., cp_size, head_dim]` + `[..., cp_size]` → `[..., head_dim]`.

`lse_mode="base2"` accepts the base-2 LSE produced by FlashInfer MLA; `lse_mode="basee"` accepts natural-log
LSE. The op does not expose `enable_pdl` because its NCCL device kernel does not use programmatic dependent launch.

- Sends each peer's partial output and LSE together, then merges received contributions in the same kernel.
- Uses NCCL device load/store-accessible (LSA) symmetric memory rather than composing separate framework collectives.
- Supports eager execution and CUDA Graph replay with a collectively created, rendezvoused workspace.
- Requires all CP ranks to be in one NCCL LSA/NVLink domain. A workspace is single-stream; concurrent streams or
  graph-capture streams need separate workspaces.

This implements the NCCL-LSA scope of #4575.

## 🔍 Related Issues

- Implements the NCCL-LSA scope of #4575.
- Related design RFC: https://github.com/NVIDIA/nccl-extensions/issues/10

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items
are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

Focused validation:

- ✅ Distributed DCP LSE-reduce test: `15 passed` per rank with 4 processes
- ✅ CPU-safe DCP API/reference tests: `12 passed`
- ✅ DCP trace signature, axes, and completeness checks: `3 passed`
- ✅ All configured pre-commit hooks on the modified PR files
- ✅ Python syntax compilation for modified Python sources
- ✅ `git diff --check`
- ✅ Four-process BF16 benchmark on four H100 80 GB GPUs in one NVLink-connected LSA domain versus unfused legacy
  A2A plus LSE merge (max-rank median; 10 warm-ups, 30 samples, 50 launches/sample):

| Shape | Fused eager | Fused + graph* | Legacy A2A + merge | Speedup |
|---|---:|---:|---:|---:|
| batch 1, heads 2, dim 64 | 12.14 µs | 10.15 µs | 178.91 µs | 14.7× |
| batch 1, heads 8, dim 128 | 15.04 µs | 13.24 µs | 179.22 µs | 11.9× |
| batch 4, heads 8, dim 128 | 18.04 µs | 16.40 µs | 178.73 µs | 9.9× |
| batch 16, heads 8, dim 128 | 18.96 µs | 17.02 µs | 180.88 µs | 9.5× |

* Fused + graph: groups 50 ops in one graph, reporting per-op average.

The largely flat legacy baseline shows that these decode-sized cases are dominated by fixed launch/collective
overhead; the speedup primarily reflects eliminating that overhead rather than a bandwidth improvement.

## Reviewer Notes

- This is the first FlashInfer op that directly uses NCCL device APIs. The new public API, AOT registration,
  trace template, documentation, tests, and benchmark are included in this PR.
- The restricted LSA/NVLink scope and the one-workspace-per-ordered-stream contract are intentional.

## Fused-kernel algorithm

For each token and attention head, every rank starts with a partial attention output $O_r$ and log-sum-exp $L_r$
from its local KV shard. The cooperative kernel completes the exchange and merge in one launch:

1. Block 0 advances a device-side epoch and selects one of two workspace slots, allowing safe slot rotation during
   CUDA Graph replay.
2. Each source rank writes its destination-specific output and LSE directly into the destination's NCCL
   symmetric-memory window using LSA/NVLink stores.
3. Sources publish system-scope release flags after their payloads are visible. Each destination acquire-waits for
   every source, then the cooperative grid synchronizes.
4. One block processes each token/head row. It computes stable rank-axis softmax weights:

$$
m = \max_r L_r, \qquad
w_r = \frac{\exp(L_r-m)}{\sum_j \exp(L_j-m)}.
$$

It then accumulates the combined output in FP32:

$$
O_{\mathrm{combined}}[d] = \sum_r w_r O_r[d].
$$

Base-2 LSE mode uses $2^{L_r-m}$. NaN and $+\infty$ LSEs are treated as $-\infty$; if every shard is empty, the
output is zero. The result is converted back to FP16 or BF16.

This removes the separate NCCL all-to-all and reduction launches by fusing peer transfer, readiness synchronization,
and the LSE-weighted merge. The implementation supports up to 64 ranks, requires one NCCL LSA/NVLink domain, and
requires each workspace to stay on one ordered CUDA stream.

## Summary by CodeRabbit

- **New Features**
  - Added an experimental context-parallel communication API that combines all-to-all exchange with
    log-sum-exp-weighted attention-output reduction.
  - Added workspace sizing and creation utilities for supported NCCL-based distributed workloads.
  - Supports natural-log and base-2 log-sum-exp calculations, including robust handling of special values.
  - Added CUDA graph and tracing support for the new operation.

- **Documentation**
  - Added API documentation covering the new communication functions and workspace requirements.

### CODE REVIEW

**Comment by kwen2501** [comment]
@aleozlx @Anerudhan This PR is ready for review. The implementation, validation results, and performance comparison are in the PR description.

**Comment by nvpohanh** [comment]
@leejnau could we add this to SGLang feature request list? @kpham-sgl  @thanhhao98 are requesting this for DCP.

This will address https://github.com/flashinfer-ai/flashinfer/issues/4575

**Comment by aleozlx** [comment]
## PR Review Screening

**CI verdict:** ✅ auto-run ok
**Review category:** `live` (rule fired: C1.2 — three net-new public names exported from `flashinfer.comm`)
**Blocking checks:** S2 (default PR template discarded), S3 (unmet), C2.1 (no hardware named for the perf claim)
**Release blocker:** no — net-new capability; no hang/crash/corruption or prior-release regression claim (C3.4)
**Early stop:** no

### Security
| Q | Answer | Evidence |
|---|--------|----------|
| S1 injection/supply-chain | no | — |
| S2 template overwritten | ❗ yes | Custom `## Summary` / `## Validation` structure; **none** of the five template headers present |
| S3 template obligations | ❗ unmet | **0 checkboxes.** The Pre-commit and Tests sections do not exist, so neither obligation is attested — though the Validation section does report a 4-process distributed run and a benchmark, so this is a structural absence rather than work not done |
| S4 agent-directing text | no | — |

### Packaging
| Q | Answer | Evidence |
|---|--------|----------|
| C1.1 external dependency bump | no | No requirements/pyproject/submodule/action changes |
| C1.2 public API changes | yes — **net-new interface** | `flashinfer/comm/__init__.py` (+7) exports `decode_cp_a2a_lse_reduce`, `decode_cp_a2a_lse_reduce_create_workspace`, and `decode_cp_a2a_lse_reduce_workspace_size`, documented in `docs/api/comm.rst`. Narrow in scope — NCCL symmetric-memory LSA only, all ranks required in one LSA domain — but it is stable public surface, not an experimental declaration (see notes) |
| C1.3 AOT registration | yes | The full checklist is present, not partially: `gen_dcp_lse_reduce_module()` in `flashinfer/jit/comm.py` (+64), registered in `flashinfer/aot.py` (+13) behind an explicit arch gate spanning sm90–sm121, a `TraceTemplate` defined in `flashinfer/trace/templates/comm.py` (+97) and wired as `@flashinfer_api(trace=decode_cp_a2a_lse_reduce_trace)`, plus the `tests/trace/template_registry.py` entry. No `fi_trace_out/` goldens — correct rather than missing: no comm template carries one |

### Presentation
| Q | Answer | Evidence |
|---|--------|----------|
| C2.1 perf claim backed by data | ❗ no | The data itself is substantial — four shapes, absolute µs for fused-eager, fused-under-graph and the legacy baseline, and the graph methodology is disclosed (50 ops per graph, per-op average). What is missing is the hardware: **no GPU model, interconnect, or node type appears anywhere** in the description or comments, so a 14.7× figure cannot be situated. The world size is given only as "four-process" |

### Implementation
| Q | Answer | Evidence |
|---|--------|----------|
| C3.1 experimental-track declared | no | No `flashinfer/experimental/` path, no `@flashinfer_experimental_api`, no `experimental` label, no tracking issue. The auto-generated release notes call it "an experimental context-parallel communication API", but that is description, not declaration — see notes |
| C3.2 shared/durable areas touched | no | **1564+/0−** — nothing is modified, only added. `csrc/dcp_lse_reduce.cu` and `include/flashinfer/comm/dcp_lse_reduce.cuh` are new files; every touch to an existing file (`comm/__init__.py`, `jit/comm.py`, `aot.py`, `trace/templates/comm.py`, `template_registry.py`, `docs/api/comm.rst`) is a registration or export |
| C3.3 tests match behavior change | yes | `tests/comm/test_dcp_lse_reduce.py` (232+) splits usefully by requirement: `test_workspace_size`, `test_workspace_size_validation` (parametrised over error cases) and `test_single_rank_reference_is_identity` (fp16/bf16 × both LSE bases) need no distribution and run anywhere; the multi-rank paths skip cleanly with "Requires torchrun, CUDA, and torch's NCCL symmetric-memory backend". Reported: 15 passed per rank across 4 processes |
| C3.4 critical fix | no | Net-new capability; no hang/crash/IMA/corruption or prior-release regression claim |

### Experimental track
| Q | Answer | Evidence |
|---|--------|----------|
| C4.1 declaration obligations | n-a | C3.1 = no |
| C4.2 machine-readable test scope | n-a | C3.1 = no |
| C4.3 isolated from common areas | n-a | C3.1 = no |

**Notes for the maintainer**
- **This is the reference shape for a new API in this repo, and worth saying so.** JIT module, AOT registration behind a real arch gate, `TraceTemplate` wired through `@flashinfer_api(trace=...)`, registry entry, docs, tests and a benchmark — all present. For contrast, flashinfer#4720 adds nine public MoE names with no trace template and eight of the nine carrying no `@flashinfer_api` at all. The gap between the two is worth pointing at when that one is reviewed.
- **Declaring the experimental track would change the routing, and may fit better.** The API is narrow by construction (NCCL-LSA only, every rank in one domain), implements a scoped part of #4575, and the release notes already describe it as experimental. Declared — tracking issue, `@flashinfer_experimental_api`, tests under `tests/experimental/` — the C1.2 waiver applies and this routes `bot` on the fast lane instead of `live`, because the track carries no semver promise. As stable surface it is a permanent commitment to an interface whose usable domain is a single NCCL LSA group.
- **The baseline in the benchmark is flat, which the speedup column hides.** "Legacy A2A + merge" reads 178.91 / 179.22 / 178.73 / 180.88 µs across shapes spanning a 16× range, while the fused path moves 12.14 → 18.96 µs. A baseline that does not respond to payload at all means the comparison is measuring fixed overhead in the legacy two-step path rather than transfer cost, so the 9.5×–14.7× band describes overhead elimination. That is a real and useful result, but it reads differently from a bandwidth win and deserves a sentence. Naming the hardware would also clear C2.1.

<sub>Generated by flashinfer-pr-screen · rubric: docs/code_review_guidance.md · not a code review · AI screening can make mistakes — a maintainer's judgment supersedes this report.</sub>

**Review by aleozlx — APPROVED** [review]

**Comment by leejnau** [comment]
@flashinfer-bot run

**Comment by kwen2501** [comment]
Hi @leejnau Previous CI was green, and I made a small change to fix the lint issue identified by pre-commit. 
Can you please let me know if the PR could be merged now? Thanks!

**Comment by kwen2501** [comment]
@flashinfer-bot run

**Comment by leejnau** [comment]
@flashinfer-bot run
